botocore/configprovider.py

```diff
@@ -851,3 +858,145 @@ def provide(self):
 
     def __repr__(self):
         return 'ConstantProvider(value=%s)' % self._value
+
+
+class ConfiguredEndpointProvider(BaseProvider):
+    """
+    NOTE: This class is considered private and is subject to abrupt breaking
+    changes or removal without prior announcement. Please do not use it
+    directly.
+    """
+
+    _ENDPOINT_URL_LOOKUP_ORDER = [
+        "environment_service",
+        "environment_global",
+        "config_service",
+        "config_global",
+    ]
+
+    def __init__(
+        self,
+        full_config,
+        scoped_config,
+        client_name,
+        environ=None,
+    ):
+        """Initialize a ConfiguredEndpointProviderChain.
+
+        :type full_config: dict
+        :param full_config: This is the dict representing the full
+            configuration file.
+
+        :type scoped_config: dict
+        :param scoped_config: This is the dict representing the configuration
+            for the current profile for the session.
+
+        :type service_name: str
+        :param service_name: service name.
+
+        :type service_id: str
+        :param service_id: service id.
+
+        :type environ: dict
+        :param environ: A mapping to use for environment variables. If this
+            is not provided it will default to use os.environ.
+        """
+        self._full_config = full_config
+        self._scoped_config = scoped_config
+        self._client_name = client_name
+
+        if environ is None:
+            environ = os.environ
+        self._environ = environ
+
+    def provide(self):
+        for location in self._ENDPOINT_URL_LOOKUP_ORDER:
+            logger.debug(
+                "Looking for endpoint for %s via: %s",
+                self._client_name,
+                location,
+            )
+
+            endpoint_url = getattr(self, f"_get_endpoint_url_{location}")()
+
+            if endpoint_url:
+                logger.info(
+                    "Found endpoint for %s via: %s.",
+                    self._client_name,
+                    location,
+                )
+                return endpoint_url
+
+        logger.debug("No configured endpoint found.")
+        return None
+
+    def _get_endpoint_url_environment_service(self):
+        return EnvironmentProvider(
+            name=self._get_service_env_var_name(), env=self._environ
+        ).provide()
+
+    def _get_endpoint_url_environment_global(self):
+        return EnvironmentProvider(
+            name="AWS_ENDPOINT_URL", env=self._environ
+        ).provide()
+
+    def unhyphenize_service_name(self, service_name):
```

Comment:
There's an analogous `hyphenize_service_id` in `utils.py` thats just as short. I moved it to `configprovider.py` since thats the only place using it, but I would be fine to remove it. The service name/service ID usage is really confusing so I thought being explicit with the intention was helpful.

Reply:
Why don't we keep it to make it explicit, but privatize it since it isn't intended to be called anywhere outside of this class.

Reply:
Looking over this, I do agree we should not make this a public method. I think @dlm6693 was just suggesting just have the method start with an underscore to explicitly note that the method is private (e.g. `_unhyphenize_service_name`). That being said I'd honestly prefer us to remove this method. See comment later on related to simplifying the logic.